internal/engine/options/options.go

```diff
@@ -0,0 +1,46 @@
+// Copyright 2024 Stacklok, Inc.
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//	http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+// Package options provides necessary interfaces and implementations
+// for implementing evaluator configuration options.
+package options
+
+import (
+	"github.com/open-feature/go-sdk/openfeature"
+
+	"github.com/stacklok/minder/internal/engine/interfaces"
+)
+
+// SupportsFlags interface advertises the fact that the implementer
+// can use an `openfeature` client to check for flags being set.
+type SupportsFlags interface {
+	SetFlagsClient(client openfeature.IClient) error
+}
+
+// Option is a function that takes an evaluator and does some
+// unspecified operation to it, returning an error in case of failure.
+type Option func(interfaces.Evaluator) error
```

Comment:
Interesting question -- do we want to have an `error` return here, or do we want options to have to be successful?

I don't have a strong feeling, but I do wonder whether we want to have the error handling propagate all over the place, or have to be handled by the Option.

Reply:
Usually, this pattern requires the developer to apply `Option` functions in some kind of constructor routine, so you'd have to handle errors anyway.

Besides, they might perform some kind of validation and the constructor should be aware of such errors.

Reply:
This is a philosophy question.  I'm willing to go with the "there might be errors, so expose error handling" philosophy, I just wanted to point out the other path ("options must not generate errors and do something sensible themselves") so we could choose.

I think we both agree that _if an option generates an error_, then the underlying object that the option was applied to is in an unspecified state and _must not be used_.

Reply:
> the other path ("options must not generate errors and do something sensible themselves")

I don't think that's useful generally, it might be very surprising to the user.
For example, we use options to configure the filter used in the `history list` endpoint. That endpoint supports filtering by time range, and requires `from` and `to` to be relatively ordered in the right way ([here](https://github.com/stacklok/minder/blob/main/internal/history/models.go#L626-L633)). Not returning an error in that case would be very, very confusing to the user, who would receive results not matching his expectation.

> I think we both agree that if an option generates an error, then the underlying object that the option was applied to is in an unspecified state and must not be used.

I totally agree on that, it might be in an inconsistent state.
I also add that, if all option succeed it's still up to the caller to verify that the object the options were applied to is in a consistent state.